File: DBD_heeje/Day_31~40/D31_169199.java
// [Programmers] 169199. 리코쳇 로봇
// 풀이 시간: 40 분

import java.util.*;

class Solution {
    public static class Point {
        int i, j;
        
        Point(int i, int j) {
            this.i = i;
            this.j = j;
        }
    }
    public static int R;
    public static int C;
    public int solution(String[] board) {
        R = board.length;
        C = board[0].length();

        // 이거 필수로 물어보기.... 코드 너무 마음에 안듦 고구마 백개 먹은 듯한 느낌
        return bfs(findRobot(board), board);
    }
    
    public Point findRobot(String[] board) {
        Point robot = null;
        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (board[i].charAt(j) == 'R') {
                    robot = new Point(i, j);
                    
                    return robot;
                }
            }
        }
        return robot;
    }
    
    public int bfs(Point robot, String[] board) {
        // python은 dir라는 예약어가 있어서 네이밍 이렇게 안하는데 java는 없나? 자동완성 안되니까 모르겠네..
        int[][] dir = { {-1, 0}, {0, -1}, {1, 0}, {0, 1} };
        int[][] v = new int[R][C];
        Queue<Point> queue = new LinkedList<>();
        queue.add(robot);
        v[robot.i][robot.j] = 1;
        
        while (!queue.isEmpty()) {
            Point point = queue.poll();
            
            if (board[point.i].charAt(point.j) == 'G') {
                return v[point.i][point.j] - 1;
            }
            
            for (int d = 0; d < 4; d++) {
                int ny = point.i + dir[d][0];
                int nx = point.j + dir[d][1];
                
                // 요거보다 밑에가 더 깔끔한 듯
                // int ny = point.i;
                // int nx = point.j;
                
                // while (inRange(ny + dir[d][0], nx + dir[d][1]) && board[ny + dir[d][0]].charAt(nx + dir[d][1]) != 'D') {
                //     ny += dir[d][0];
                //     nx += dir[d][1];
                // }                
                
                while (inRange(ny, nx) && board[ny].charAt(nx) != 'D') {
                    ny += dir[d][0];
                    nx += dir[d][1];
                }
                
                // 대신 더한 것 하나 빼줘야 한다는 아쉬움..
                ny -= dir[d][0];
                nx -= dir[d][1];
                
                if (v[ny][nx] == 0) {
                    // Point 쓰니까 변수명이 살짝 마음에 안드는데??
                    // 원래는 v[ny][nx] = v[y][x] + 1 요렇게 썼었는뎅..
                    v[ny][nx] = v[point.i][point.j] + 1;
                    queue.add(new Point(ny, nx));
                }
            }   
        }
        
    return -1;
    }
    
    public boolean inRange(int y, int x) {
        return y >= 0 && y < R && x >= 0 && x < C;
    }
}

File: DBD_kyungmin/Day_31~40/D31_169199.py
# [Programmers] 169199. 리코쳇 로봇
# 풀이 시간: 70 분

from collections import deque

# bfs
def solution(board):
    # 우 하 좌 상
    dr = [0, 1, 0, -1]
    dc = [1, 0, -1, 0]
    
    N = len(board)
    M = len(board[0])
    
    # bfs 함수 
    def bfs(sr, sc):
        # 큐 초기화 
        q = deque()
        # 방문 배열 초기화
        # 방문 배열에 이동 횟수 카운트 
        visited = [[0] * M for _ in range(N)]
        
        q.append((sr, sc))
        # 시작 지점은 이동 횟수가 0이어야 하지만, 시작 지점의 방문 처리를 위해 1로 설정
        visited[sr][sc] = 1
        
        while q:
            # 큐에서 하나 빼기 
            cr, cc = q.popleft()
            
            # 종료 지점이라면 
            if board[cr][cc] == 'G':
#                 for k in visited:
#                     print(k)
                # 시작을 1로 시작했기 때문에 1 빼주기
                return visited[cr][cc] - 1
            
            # 사방 탐색
            for dir in range(4):
                nr = cr
                nc = cc
                
                # 현재에서부터 한번 쭉 직진할 수 있는 끝 지점만 찾기
                # 끝 지점만 큐에 넣기
                # (직진하는 중간 지점들을 넣으면, 그 지점들에 대한 사방 탐색이 이루어지기 때문)
                
                # 조건에 맞는다면 현재 방향으로 쭉 직진
                while True:
                    # 한 칸 이동
                    nr += dr[dir]
                    nc += dc[dir]
                                        
                    # 1. 탐색 지점이 이동할 수 있는 곳인가? 
                    # 2. 탐색 지점이 이동할 수 없는 곳인가?: 탐색 지점이 범위 외거나, D인 경우 
                    # 3. 탐색 지점이 (직진)이동을 끝내는 곳인가?: while문 밖(쭉 직진할 경로를 뚫은 상태)
                    # => 2번과 3번의 구분이 중요함 
                    
                    # 조건 확인
                    # 범위 외
                    if 0 > nr or nr >= N or 0 > nc or nc >= M:
                        # 이동 안 함, 이동 되돌리기 
                        nr -= dr[dir]
                        nc -= dc[dir]
                        
                        break
                    # 장애물
                    elif 0 <= nr < N and 0 <= nc < M and board[nr][nc] == 'D':
                        # 이동 안 함, 이동 되돌리기 
                        nr -= dr[dir]
                        nc -= dc[dir]
                        
                        break
                        
                    # 이동 안 한다면 nr nc는 한 칸 이전의 지점이 되면서 while 탈출 => 이 지점이 방문하지 않은 지점이라면 끝 지점이 됨
                    
                # 직진 끝       
                # 미방문한 곳이라면
                if not visited[nr][nc]:
                    # 이동 
                    q.append((nr, nc))
                    visited[nr][nc] = visited[cr][cc] + 1

        return -1                    
    
    for i in range(N):
        for j in range(M):
            if board[i][j] == 'R':
                return bfs(i, j)